asmcli/lib/parse.sh

```diff
@@ -0,0 +1,390 @@
+        shift 2
+        ;;
+      -D | --output_dir | --output-dir)
+        arg_required "${@}"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "OUTPUT_DIR" "${2}")"
+        shift 2
+        ;;
+      --dry_run | --dry-run)
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "DRY_RUN" 1)"
+        shift 1
+        ;;
+      --only_validate | --only-validate)
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "ONLY_VALIDATE" 1)"
+        shift 1
+        ;;
+      --only_enable | --only-enable)
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "ONLY_ENABLE" 1)"
+        shift 1
+        ;;
+      --ca_cert | --ca-cert)
+        arg_required "${@}"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CA_CERT" "${2}")"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CUSTOM_CA" 1)"
+        shift 2
+        ;;
+      --ca_key | --ca-key)
+        arg_required "${@}"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CA_KEY" "${2}")"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CUSTOM_CA" 1)"
+        shift 2
+        ;;
+      --root_cert | --root-cert)
+        arg_required "${@}"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CA_ROOT" "${2}")"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CUSTOM_CA" 1)"
+        shift 2
+        ;;
+      --cert_chain | --cert-chain)
+        arg_required "${@}"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CA_CHAIN" "${2}")"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CUSTOM_CA" 1)"
+        shift 2
+        ;;
+      -r | --revision_name | --revision-name)
+        arg_required "${@}"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "REVISION_LABEL" "${2}")"
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "CUSTOM_REVISION" 1)"
+        shift 2
+        ;;
+      -v | --verbose)
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "VERBOSE" 1)"
+        shift 1
+        ;;
+      -h | --help)
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "PRINT_HELP" 1)"
+        shift 1
+        ;;
+      --version)
+        CONTEXT_JSON="$(context_set-option-from-variable "${CONTEXT_JSON}" "PRINT_VERSION" 1)"
+        shift 1
+        ;;
+      *)
+        fatal_with_usage "Unknown option ${1}"
+        ;;
+    esac
+  done
+
+  # make sure to redirect stdout as soon as possible if we're dumping the config
+  local PRINT_CONFIG; PRINT_CONFIG="$(context_get-option-from-variable "${CONTEXT_JSON}" "PRINT_CONFIG")"
+  if [[ "${PRINT_CONFIG}" -eq 1 ]]; then
+    exec 3>&1
+    exec 1>&2
+  fi
+}
+
+validate_args() {
```

Comment:
High level comment: I'm imagining that specific components will be responsible for some of these checks. For example, MCP should be responsible for validating that there aren't any custom overlays passed in, and this function should mostly be concerned with validating flags that apply to multiple components or subcommands (like "you can't use --print-config and --enable at the same time").

Nothing needs to necessarily change for this PR, but I wanted to make sure we didn't spend too much time testing or worrying about the format here since it will shrink as more files get split out.

Reply:
Ack